pkg/engine/lifecycle_test.go

```diff
@@ -2970,3 +2969,206 @@ func TestSingleResourceIgnoreChanges(t *testing.T) {
 		"b": resource.NewNumberProperty(4),
 	}, []string{"a"}, []deploy.StepOp{deploy.OpUpdate})
 }
+
+// TestDefaultProviderDiff tests that the engine can gracefully recover whenever a resource's default provider changes
+// and there is no diff in the provider's inputs.
+func TestDefaultProviderDiff(t *testing.T) {
+	const resName, resBName = "resA", "resB"
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("0.17.10"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{}, nil
+		}),
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("0.17.11"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{}, nil
+		}),
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("0.17.12"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{}, nil
+		}),
+	}
+
+	runProgram := func(base *deploy.Snapshot, versionA, versionB string, expectedStep deploy.StepOp) *deploy.Snapshot {
+		program := deploytest.NewLanguageRuntime(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+			_, _, _, err := monitor.RegisterResource("pkgA:m:typA", resName, true, "", false, nil, "",
+				resource.PropertyMap{}, nil, false, versionA, nil)
+			assert.NoError(t, err)
+			_, _, _, err = monitor.RegisterResource("pkgA:m:typA", resBName, true, "", false, nil, "",
+				resource.PropertyMap{}, nil, false, versionB, nil)
+			assert.NoError(t, err)
+			return nil
+		})
+		host := deploytest.NewPluginHost(nil, nil, program, loaders...)
+		p := &TestPlan{
+			Options: UpdateOptions{host: host},
+			Steps: []TestStep{
+				{
+					Op: Update,
+					Validate: func(project workspace.Project, target deploy.Target, j *Journal,
+						events []Event, res result.Result) result.Result {
+						for _, entry := range j.Entries {
+							if entry.Kind != JournalEntrySuccess {
+								continue
+							}
+
+							switch entry.Step.URN().Name().String() {
+							case resName, resBName:
+								assert.Equal(t, expectedStep, entry.Step.Op())
+							}
+						}
+						return res
+					},
+				},
+			},
+		}
+		return p.Run(t, base)
+	}
+
+	// This test simulates the upgrade scenario of old-style default providers to new-style versioned default providers.
+	//
+	// The first update creates a stack using a language host that does not report a version to the engine. As a result,
```

Comment:
These comments are all slightly misleading after your refactoring to create two resources (with possible different versions) FWIW.